packages/core/js-client/src/api.ts

```diff
@@ -15,182 +15,158 @@
+        }
 
-  return {
-    client: peer,
-    args: argsRes,
-    config: config,
+        const wrappedFunction = wrapFunction(arg, argSchema);
+
+        return [argNames[i], wrappedFunction];
+      }
+
+      if (typeof arg === "function") {
+        throw new Error("Argument and schema doesn't match");
+      }
+
+      return [argNames[i], ts2aqua(arg, argSchema)];
+    }),
+  );
+
+  const returnTypeVoid =
+    def.arrow.codomain.tag === "nil" || def.arrow.codomain.items.length === 0;
+
+  const params = {
+    peer,
+    args: callArgs,
+    config,
   };
-}
+
+  const result = await callAquaFunction({
+    script,
+    ...params,
+    fireAndForget: returnTypeVoid,
+  });
+
+  const valueSchema =
+    def.arrow.codomain.tag === "unlabeledProduct" &&
+    def.arrow.codomain.items.length === 1
+      ? def.arrow.codomain.items[0]
+      : def.arrow.codomain;
+
+  return aqua2ts(result, valueSchema);
+};
 
 /**
- * Arguments could be passed in one these configurations:
- * [serviceObject]
- * [peer, serviceObject]
- * [defaultId, serviceObject]
- * [peer, defaultId, serviceObject]
- *
- * Where serviceObject is the raw object with function definitions passed by user
- *
- * This function select the appropriate configuration and returns
- * arguments in a structured way of: { peer, serviceId, service }
+ * Convenience function to support Aqua `service` generation backend
+ * The compiler only need to generate a call the function and provide the corresponding definitions and the air script
+ * @param args - raw arguments passed by user to the generated function
+ * @param def - service definition generated by the Aqua compiler
  */
-const extractServiceArgs = (
-  args: unknown[],
-  defaultServiceId?: string,
-): {
-  peer: FluencePeer;
-  serviceId: string | undefined;
-} => {
-  let peer: FluencePeer;
-  let serviceId: string | undefined;
-
-  if (isFluencePeer(args[0])) {
+export const v5_registerService = (args: unknown[], def: ServiceDef): void => {
+  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+  const serviceImpl = args.pop() as ServiceImpl;
+  let peer: FluencePeer | undefined;
+  let serviceId = def.defaultServiceId;
+
+  if (args[0] instanceof FluencePeer) {
     peer = args[0];
     args = args.slice(1);
   } else {
-    if (Fluence.defaultClient == null) {
-      throw new Error(
-        "Could not register Aqua service because the client is not initialized. Did you forget to call Fluence.connect()?",
-      );
-    }
-
     peer = Fluence.defaultClient;
   }
 
+  if (peer == null) {
+    throw new Error(
+      "Could not register Aqua service because the client is not initialized. Did you forget to call Fluence.connect()?",
+    );
+  }
+
   if (typeof args[0] === "string") {
     serviceId = args[0];
-  } else {
-    serviceId = defaultServiceId;
   }
 
-  return {
+  if (serviceId == null) {
+    throw new Error("Service ID is not provided");
```

Comment:
It might be provided, but be of a wrong type, eg accidentally wrapped in option.

How to improve this error so the user understands what to fix?

Reply:
Thanks, refactored